astro/src/content/blog/hundreds-millions-entities.mdx

```diff
@@ -0,0 +1,261 @@
+---
+publish_date: 2024-10-23
+title: Hundreds Of Millions Of Entities
+description: How Does FusionAuth Handle Large Numbers Of Entities?
+authors: Dan Moore
+image: /img/blogs/hundreds-millions-entities/hundreds-of-millions-of-entities.png
+categories: Education
+tags: performance, entities, load testing, clownwear
+excerpt_separator: "{/* more */}"
+---
+import Breadcrumb from 'src/components/Breadcrumb.astro';
+import Aside from 'src/components/Aside.astro';
+
+Recently we load tested a FusionAuth Cloud deployment to see how it handled large numbers of entities. The goal was to have hundreds of millions of entities, associate users with them, and see if the system still performed well with and without pushing entity data into a token during a login.
+
+{/* more */}
+
+## The Use Case
+
+This large number of entities comes up in a business-to-business-to-business (B2B2B) scenario where:
+
+* the FusionAuth customer has customers have their own customers
+* users can belong to one or many companies
+* users can have different permissions for different customers.
+
+That's a bit abstract, so let's outline a scenario.
+
+Cosmo's Clown Store, which [previously added passkeys](/articles/authentication/webauthn-explained), has grown so large it is going to franchise. Within each franchise, there are employees, stores and departments. Each employee will have some relationship to a store (such as being a worker or a manager) and possibly a relationship to a department (purchasing, HR, marketing). The latter only occurs when a franchise has more than two stores. There also might be a number of Cosmo's Clown Store customers who would be users without any such relationship.
+
+As part of the franchise agreement, Cosmo's Clown Store offers multiple applications to help franchisees manage their stores, including Clownwear Central (an ERP system), Tell It To The Nose (a community management tool), and Juggling Your Hours (a timesheet application). All employees have varying levels of access to these tools, and an employee might be a manager at one store with ERP access, and part time help at another store, with only Juggling Your Hours access.
+
+Relationships like these can be modeled with entities, and FusionAuth can handle a large amount of them efficiently.
+
+## The Deployment
+
+The deployment that was tested is an HA (High Availability) running three extra large nodes.
```

Comment:
We should start to make an effort to eliminate passive voice, in line with the original Inversoft style guide. I don't think it's a big enough deal to add a change here, but it's something that I'm trying to focus on moving forward.

Reply:
We can use vale to check for that, though it isn't perfect. I fixed this one.